File: Typescript/HeapSort.ts
var array_length : number;
/* to create MAX  array */
function heap_root(input : number[], i : number) {
   var left = 2 * i + 1;
   var right = 2 * i + 2;
   var max = i;

   if (left < array_length && input[left] > input[max]) {
       max = left;
   }

   if (right < array_length && input[right] > input[max])     {
       max = right;
   }

   if (max != i) {
       swap(input, i, max);
       heap_root(input, max);
   }
}

function swap(input : number[], index_A : number, index_B : number) {
   var temp = input[index_A];

   input[index_A] = input[index_B];
   input[index_B] = temp;
}

function heapSort(input : number[]) : number[] {

   array_length = input.length;

   for (var i = Math.floor(array_length / 2); i >= 0; i -= 1)      {
       heap_root(input, i);
     }

   for (i = input.length - 1; i > 0; i--) {
       swap(input, 0, i);
       array_length--;


       heap_root(input, 0);
   }

   return input;
}

export default heapSort;

File: C/heapSort.c
#include<stdio.h>
void heapify(int arr[],int n,int i){
	int root,left,right,tmp;
	root = i;
	left = 2 * i + 1;
	right = 2 * i + 2;
	if(left < n && arr[left] > arr[root])
		root = left;
	if(right < n && arr[right] > arr[root])
		root = right;
		
	if(root != i){
		tmp = arr[root];
		arr[root] = arr[i];
		arr[i] = tmp;
		heapify(arr,n,root);
	}
} 
void heapsort(int arr[],int n){
	int i,tmp;
	for(i = (n/2 - 1); i >= 0; i--){
		heapify(arr,n,i);
	}
	for(i = n-1; i >= 0; i--){
		tmp = arr[0];
		arr[0] = arr[i];
		arr[i] = tmp;
		heapify(arr,i,0);
	}
}
int main(){
    
	int i,n;
	printf("Enter number elements to be sorted:" );
    scanf("%d",&n);
    int arr[n];
	for(i = 0; i < 5; i++)
		scanf("%d",&arr[i]);
	heapsort(arr,n);
	for(i = 0; i < n; i++) 
		printf("%d ",arr[i]);
return 0;
}
